SwiftCompilerSources/Sources/Optimizer/Utilities/EscapeInfoWalker.swift

```diff
@@ -0,0 +1,834 @@
+  ///    %ref1 = struct_extract %value, #field1      // ignored - not selected by path
+  ///    %addr = ref_element_addr %ref, #some_field  // path is empty
+  ///    %v = load %addr                             // path is empty
+  ///    return %v                                   // escaping!
+  /// \endcode
+  ///
+  /// Trivial values are ignored, even if they are selected by `path`.
+  mutating func isEscaping(object: Value, path: Path = Path(),
+                           visitor: EscapeInfoWalkerVisitor = DefaultEscapeInfoWalkerVisitor()) -> Bool {
+    start(forAnalyzingAddresses: false, visitor: visitor)
+    defer { cleanup() }
+    
+    let state = State(followStores: false, knownType: nil)
+    if let (path, state) = shouldRecomputeUp(def: object, path: path, state: state) {
+      if object.type.isAddress {
+        return walkUp(address: object, path: path, state: state)
+      } else {
+        return walkUp(value: object, path: path, state: state)
+      }
+    }
+    return false
+  }
+  
+  /// Returns true if the definition of `value` is escaping.
+  ///
+  /// In contrast to `isEscaping`, this function starts with a walk-down instead of a walk-up from `value`.
+  mutating func isEscapingWhenWalkingDown(object: Value, path: SmallProjectionPath = SmallProjectionPath(),
+                                          visitor: EscapeInfoWalkerVisitor = DefaultEscapeInfoWalkerVisitor()) -> Bool {
+    start(forAnalyzingAddresses: false, visitor: visitor)
+    defer { cleanup() }
+    
+    return cachedWalkDown(def: object, path: path, state: State(followStores: false, knownType: nil))
+  }
+  
+  /// Returns true if any address of `value`, which is selected by `path`, can escape.
+  ///
+  /// For example, let's assume this function is called with a struct, containing a reference,
+  /// and a path of `s0.c*.v**`:
+  /// \code
+  ///    %value : $Struct<X>                         // path == s0.c*.v**, the initial `value`
+  ///    %ref = struct_extract %value, #field0       // path == c*.v**
+  ///    %selected_addr = ref_element_addr %ref, #x  // path == v**,       the selected address
+  ///    apply %f(%selected_addr)                    // escaping!
+  /// \endcode
+  ///
+  /// There are two differences to `isEscaping(object:)`:
+  /// * Loads from the selected address(es) are ignored. So it's really about the _address_ and
+  ///   not the value stored at the address.
+  /// * Addresses with trivial types are _not_ ignored.
+  mutating
+  func isEscaping(addressesOf value: Value, path: SmallProjectionPath = SmallProjectionPath(.anyValueFields),
+                  visitor: EscapeInfoWalkerVisitor = DefaultEscapeInfoWalkerVisitor()) -> Bool {
+    start(forAnalyzingAddresses: true, visitor: visitor)
+    defer { cleanup() }
+    
+    let state = State(followStores: false, knownType: nil)
+    if let (path, state) = shouldRecomputeUp(def: value, path: path, state: state) {
+      if value.type.isAddress {
```

Comment:
This `if value.type.isAddress { walkUp(address...) } else { walkUp(value:...) }` pattern appears a lot in this file. It's probably worth extracting it to a small utility function: `func walkUp(addressOrValue:...)`